### PR TITLE
Dir state vault

### PR DESCRIPTION
**Pull request by jstefl**
At first, I was thinking to handle directories within `FileStateVault` but it turned out, that would be a mess. Main problem what I saw was in conflicting records (what with files in already pushed directories). It could be implementable of course, but it would had to be documented well and I am not convinced by clearness/cleanliness of such a solution. 

For 'natural semantic sake', I handle directories in separate class.

### CODE REVIEW

**Comment by m-czernek** [comment]
The code looks nice; I only wonder, does it handle permissions/ownerships in any way? I.e. if the directory is owned by Apache, it is recreated as being owned by the process owner, right? It is possible we don't mind that, only that we cannot use it as a total replacement for copying and restoring all confs.

**Comment by jstefl** [comment]
Hm, good comment - do we delete directories *in* configurations directories ever? Since imho, this is the only case when this has effect (we don't delete whole configuration directories like con.f). Yes, it could happen, but is it real use case what we use or we are going to use? To my best knowledge we modify just files, we don't delete directories from httpd configuration dirs. Additionally, I checked last jbcs/httpd/zip and there are no directories within configuration directories, do we have it somewhere else?

Imho, *if* smbd. want to delete con.d they can handle it as the they need it in this exceptional situation.

I think I can implement the support, but I wouldn't like to do that now just for sake of theory tbh. 

P.S.: Btw. this feature does not have `FileStateVault` either, so when is any file deleted it is not recovered with previous access rights.

**Comment by m-czernek** [comment]
I don't think we delete any directories, hence why I think we probably don't mind this. I was more looking ahead to using this for replacing the backupAll and restoreAll methods with something more fine-grained, for which usecase the Vault classes aren't really great. However, implementing support for user, group, and perms would be non-trivial I'd guess.

Overall, I'm for merging this, only that it's a limitation that's good to keep in mind.

**Comment by jan-onderka** [comment]
@jstefl It is not just theory. We are delaying BaseOS httpd MR #48 in sake of better implementation of file/dir state vault to backup and restore all httpd configuration folders with their content. So in this case, I need full support for file/directory ownership and permissions.

**Comment by jstefl** [comment]
@jan-onderka in that case, please do no wait to this MR.

**Comment by jstefl** [comment]
I have added support for saving and reverting permissions. 
I checked that it works manually, I will write some unit tests probably tomorrow.

**Comment by jstefl** [comment]
I have added support for handling targets what are not accessible during push operations.

Still not mergable because of missing relevant unit tests and possible some minor changes.

**Comment by m-czernek** [comment]
Changing the title to WIP; when the PR is ready, just remove the WIP flag please.

**Comment by jstefl** [comment]
Ready for review.

**Comment by m-czernek** [comment]
The changes look great to me, it's really nice that this PR is using JBFile's perms functionality. @jstefl can you just check the comments I made before?

File: core/src/main/groovy/noe/common/utils/DirStateVault.groovy [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+   * If directory toStore does exist it's current content is stored into memory.
+   * If does not then this fact is just noted.
+   *
+   * All subdirectories are handled recursively.
+   *
+   * The state of the file can be loaded by `pop(...)` or `popAll()` methods, they restore
+   * last saved state and delete this state from memory.
+   */
+  @Override
+  DirStateVault push(File toStore) {
```
[the remark this answers is not in the document]
Comment (jstefl) [reply core/src/main/groovy/noe/common/utils/DirStateVault.groovy]: Fixed in commit "FileStateVault: Type, DirStateVault: Exception on pushing file".

File: core/src/main/groovy/noe/common/utils/FileStateVault.groovy [hunk 2]
```diff
@@ -13,10 +14,12 @@ package noe.common.utils
  *
  * There is no manipulation with access rights on files, only exception is when file
  * is being removed - when normal delete fails deleting with sudo is activated (if enabled).
+ *
  * @see JBFile#delete
+ * @see DirStateVault
  *
  */
-class FileStateVault {
+class FileStateVault implements StateVault {
```
[the remark this answers is not in the document]
Comment (jstefl) [reply core/src/main/groovy/noe/common/utils/FileStateVault.groovy]: Fixed in commit "FileStateVault: Type, DirStateVault: Exception on pushing file".

**Comment by jstefl** [comment]
All comments incorporated.

**Comment by m-czernek** [comment]
I have no further comments. The code looks nice. Thank you, merging.

**Pull request merged**